### PR TITLE
Fixed counter on failed connection

### PR DESCRIPTION
**Pull request by qweryty**
Sometimes `self._connection.acquire()` can throw an exception and `self._connection_counter` will remain 1, in that case, if we try to reconnect `__aenter__` will always skip `self._connection.acquire()` and we won't be able to make any further requests to database.

### CODE REVIEW

**Comment by taybin** [comment]
I believe I have been encountering this issue, so it would be great to see this or something similar to it merged.

**Comment by taybin** [comment]
Looks like it  just needs a unittest.

File: databases/core.py [hunk 1]
```diff
@@ -237,8 +237,12 @@ def __init__(self, backend: DatabaseBackend) -> None:
     async def __aenter__(self) -> "Connection":
         async with self._connection_lock:
             self._connection_counter += 1
```
Comment (aminalaee) [inline databases/core.py:239]:
```suggestion
            if self._connection_counter == 0:
                await self._connection.acquire()
                self._connection_counter += 1
```

**Review by aminalaee — COMMENTED** [review]
Maybe this can be simplified a bit?

File: databases/core.py (hunk 1 above)
Comment (aminalaee) [reply databases/core.py:239]: If any exception happens it won't be incremented at all, so we don't have to decrement in the except part.

Comment (aminalaee) [reply databases/core.py:239]: @taybin @qwerty what do you think? This is simpler and won't need the extra test,

Comment (taybin) [reply databases/core.py:239]: Does this still work with concurrent requests?  It removes the semaphore.  In which case it doesn't even need the counter.  The connection lock would be enough.

But is the connection lock enough or do we need the semaphore?

Comment (aminalaee) [reply databases/core.py:239]: @taybin I'm not really sure, I'll try a test, if it doesn't work out, we'll merge it as is.

**Comment by aminalaee** [comment]
I'll try to add the test for new code, @taybin feel free to add that if you like.

**Comment by taybin** [comment]
I'm working on a unittest for this right now.

**Comment by taybin** [comment]
@aminalaee I have a PR with @qweryty's code and a unittest at #385

**Pull request closed without merging**